spinedb_api/db_mapping_base.py

```diff
@@ -494,19 +539,76 @@ def add_item_from_db(self, item):
         Returns:
             tuple(MappedItem,bool): The mapped item and whether it hadn't been added before.
         """
-        current = self.find_item_by_id(item["id"], fetch=False) or self._find_item_by_unique_key(
-            item, fetch=False, complete=self._db_map.has_external_commits()
-        )
-        if current:
-            if current.status == Status.to_add:
-                current["id"].resolve(item["id"])
-                current.status = Status.committed
-            return current, False
-        item = self._make_and_add_item(item)
+        mapped_item = self._find_item_by_unique_key(item, fetch=False, valid_only=False)
+        if mapped_item:
+            self._force_id(mapped_item, item["id"])
+            return mapped_item, False
+        mapped_item = self.get(item["id"])
+        if mapped_item is not None and mapped_item.is_equal_in_db(item):
+            return mapped_item, False
+        self._free_id(item["id"])
+        mapped_item = self._make_and_add_item(item)
         if self.purged:
             # Lazy purge: instead of fetching all at purge time, we purge stuff as it comes.
-            item.cascade_remove(source=self.wildcard_item)
-        return item, True
+            mapped_item.cascade_remove(source=self.wildcard_item)
+        return mapped_item, True
+
+    def _force_id(self, mapped_item, id_):
+        """Makes sure that mapped_item has the given id_, corresponding to the new id of the item
+        in the DB after some external changes.
+
+        Args:
+            mapped_item (MappedItemBase): An item in the in-memory item.
+            id_ (int): The most recent id_ of the item as fetched from the DB.
+        """
+        mapped_id = mapped_item["id"]
+        if mapped_id == id_:  # This is True even if mapped_id is a TempId resolved to the given id_
+            return
```

Comment:
Is it correct that, when coming from `_make_and_add_item()`, we do not check here if the db item is actually same item as the mapped item, e.g. it may have a different description? Could this lead to a situation where the mapped items differ from database items and there is no way of "seeing" or committing those differences?

Reply:
That's right, I missed that and @jkiviluo pointed out in element. Those are *user*-data conflicts, we just resolve them by favoring the mapped-item. This PR is only about *internal*-data conflicts, aka id conflicts (e.g., Widget and Gadget exchanging their ids after running an external script).

User-data conflict resolution should be done in a different PR in my opinion?

Reply:
Right, I got the impression that we handle the user data conflicts already in this PR from your comment in Element so thanks for clearing that confusion. I guess it is orthogonal to the problem we have at hand and can be done in another PR. Better take small steps here.